File: 345-reverseVowelsOfString/solution.py
class Solution(object):
    def reverseVowels(self, s):
        """
        :type s: str
        :rtype: str
        """

        # 30% O(n) for indexing
        vowels = ['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U']
        # 80%             important!!!   O(1) for indexing
        vowels = set(['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'])

        indice = []
        for i in range(len(s)):
            if s[i] in vowels:
                indice.append(i)
        ls = list(s)
        for j in range(int(len(indice)/2)):
            temp = ls[indice[j]]
            
            ls[indice[j]] = ls[indice[-j-1]]
            ls[indice[-j-1]] = temp
            
        return "".join(ls)


        # 10%
        st=[]
        ls = list(s)
        for i in range(len(ls)):
            if ls[i] in vowels:
                st.append(ls[i])
        
        for i in range(len(ls)):
            if ls[i] in vowels:
                ls[i] = st.pop()
        
        
        return "".join(ls)

File: 345-reverseVowelsOfString/solution.java
public class Solution {
    public String reverseVowels(String s) {
        // 54%
        Set<Character> vowels = new HashSet<>(Arrays.asList('a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'));
        char[] chars = s.toCharArray();
        
        int l = 0, r = chars.length-1;
        while (l < r) {
            while (l < r && !vowels.contains(chars[l])) {
                l++;
            }
            while (l < r && !vowels.contains(chars[r])) {
                r--;
            }
            char tmp = chars[l];
            chars[l++] = chars[r];
            chars[r--] = tmp;
        }
        return new String(chars);
        

        // 10%
        Stack<Character> st = new Stack<Character>();
        char[] chars = s.toCharArray();
        for (char c : chars){
            if (vowels.contains(Character.toLowerCase(c))){
                st.push(c);
            }
        }
        
        for (int i=0; i< chars.length; i++){
            if (vowels.contains(Character.toLowerCase(chars[i]))){
                chars[i] = st.pop();
            }
        }

        return new String(chars);
    }
}